### ISSUE TITLE
_adapter_disconnect is not called properly and the socket remain enabled for select adapaters

### ISSUE THREAD

**Issue by pfreixes** [pika/pika#634]
I came across with this bug when I saw duplicated connections, or more, in my rabbitmq sever using the `rabbitmqctl list_connections` command where these connections belong to the same Pika AMQP process.

After spend a lot of time guessing which could be the origin of the bug, at last I reproduced it and fixed calling by hand the `_adapter_disconnect` connection method 

```python
    def _on_connection_closed(self, connection_unused, reply_code, reply_text):
        logging.warning('{0} Server closed connection, reopening: ({1}) {2}'.format(self.host,
                                                                                    reply_code,
                                                                                    reply_text))

        # PIKA doesn't call the remove handler of current fd belonging to the socket when
        # there is a communication issue between the server and client, therefore the socket
        # remains open and future heartbeats sent to outdated connection can provoke duplicate
        # duplicated conections scenarios. We call it by hand to avoid that
        connection_unused._adapter_disconnect()

        # after here some strategy to reconnect the client
```

The connection is configured to use the heartbeat strategy to figure out when there are communications issues, to allow to the client runs a strategy to connect again after N seconds. 

```python
    @gen.coroutine
    def _connect(self):
        coroutiner = CoroutineMaker({'TornadoConnection': 'on_open_callback'})
        try:
            credentials = pika.PlainCredentials(self.username, self.password)
            connection_parameters = pika.ConnectionParameters(
                host=self.host,
                credentials=credentials,
                connection_attempts=1,#  reconnection strategy is made by our own
                heartbeat_interval=60)
            self._connection = yield coroutiner.convert(pika.adapters.TornadoConnection)(
                connection_parameters,
                on_open_error_callback=self._on_connection_error)

```

The issue comes when the communications between the server and the client is stalled because of network issues. When the heartbeat module tries to close the connection [1] it calls first of all the method to close the connection and after it the `on_disconnect` method that will fire our callback.

The `connection.close` method tries to close the connection using a set of good practices, such as close the channels, send the properly Close frame, etc. And at last delegating to the callback , ,previously registered to attend the response of the Close frame, some other stuff like call the `_adapter_disconnect` method.

As we saw before the scenario can not grantee that the Close frame is delivered and the most important that the response of this frame is always received.

In that situation the map of the file descriptors that handles the Tornado adapter has still, because the unregistered was called never, the file descriptor belonging to one connection that has been apparently closed.

Because of that, if the network issues vanish the Rabbitmq sever, or SO buffers, can still send data to the Pika AMQP process. By default Pika will show the next error because of this data:

```
2015-08-07 16:39:19,952 - pika.adapters.base_connection - ERROR - Read empty data, calling disconnect
2015-08-07 16:39:19,952 - pika.adapters.base_connection - WARNING - Connection is closed but not stopping IOLoop
2015-08-07 16:39:19,953 - pika.connection - WARNING - Disconnected from RabbitMQ at manager.miimetiq.m2mcloudfactory.com:5672 (0): Not specified
```

And pika will call again the `_on_connection_closed` when in fact the connection is open and running.

[1] https://github.com/pika/pika/blob/0.9.14/pika/heartbeat.py#L118

**Comment by gmr** [comment]
Thanks for the detailed analysis. Have you verified this is an issue in 0.10.0/master?

**Comment by vitaly-krugl** [comment]
@pfreixes: We took care of some of the connection cleanup issues in the 0.10.0 pre-release.

@gmr: Looking at the implementation of `HeartbeatChecker._close_connection` below, it looks like there are a few potential problems there as well:

1. It doesn't take continuation-style adapters into account: when it calls `self._connection.close(...)`, that operation will not complete in continuation-style adapters before `self._connection.close` returns. This is because the graceful close operation needs to send message(s) to the broker and wait for response(s), which won't complete in the context of that function.
2.  Since the heartbeat checker has determined that there is a networking issue, there is probably no value in attempting to perfrom the graceful connection close operation.
3. The call to the private method `self._connection._on_disconnect` seems out of place here. There should probably be a reasonable code path in the `Connection` class that's responsible for this chunk of the adapter's business logic.

```
    def _close_connection(self):
        """Close the connection with the AMQP Connection-Forced value."""
        LOGGER.info('Connection is idle, %i stale byte intervals',
                    self._idle_byte_intervals)
        duration = self._max_idle_count * self._interval
        text = HeartbeatChecker._STALE_CONNECTION % duration
        self._connection.close(HeartbeatChecker._CONNECTION_FORCED, text)
        self._connection._adapter_disconnect()
        self._connection._on_disconnect(HeartbeatChecker._CONNECTION_FORCED,
                                        text)
```

**Comment by vitaly-krugl** [comment]
@pfreixes and @gmr: The closing of the connection in this scenario should be addressed by PR #685
